File: 2023/src/Day08.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * @author Zachary Cockshutt
 * @since  2023-12-09
 */
public class Day08 extends Day
{
    @Override
    public String partOne() throws IOException
    {
        int count = 0;
        var input = this.input();
        var dirs = input[0].toCharArray();
        var nodes = new HashMap<String, Node>();
        for (int i = 2; i < input.length; i++)
        {
            var codes = input[i].replace(")","").split(" = \\(|, ");
            var node = new Node(codes[1], codes[2]);
            nodes.put(codes[0], node);
        }
        var curr = "AAA";
        while (!curr.equals("ZZZ"))
        {
            int i = Math.floorMod(count++, dirs.length);
            var node = nodes.get(curr);
            var dir = dirs[i];
            curr = dir == 'L' ? node.left : node.right;
        }
        return String.valueOf(count);
    }

    @Override
    public String partTwo() throws IOException
    {
        var input = this.input();
        var dirs = input[0].toCharArray();
        var nodes = new HashMap<String, Node>();
        var starts = new ArrayList<String>();
        for (int i = 2; i < input.length; i++)
        {
            var codes = input[i].replace(")","").split(" = \\(|, ");
            var node = new Node(codes[1], codes[2]);
            nodes.put(codes[0], node);
            if (codes[0].charAt(2) == 'A') { starts.add(codes[0]); }
        }
        var counts = new int[starts.size()];
        for (int i = 0; i < counts.length; i++)
        {
            var curr = starts.get(i);
            while (curr.charAt(2) != 'Z')
            {
                int j = Math.floorMod(counts[i]++, dirs.length);
                var node = nodes.get(curr);
                var dir = dirs[j];
                curr = dir == 'L' ? node.left : node.right;
            }
        }
        return String.valueOf(lcm(counts));
    }

    public static long lcm(int[] arr)
    {
        long lcm = 1;
        int divisor = 2;
        while (true) 
        {
            int count = 0;
            boolean divisible = false;
            for (int i = 0; i < arr.length; i++) 
            {
                if (arr[i] == 0)     { return 0; }
                else if (arr[i] < 0) { arr[i] = arr[i] * (-1); }
                if (arr[i] == 1) { count++; }
                if (arr[i] % divisor == 0) 
                {
                    divisible = true;
                    arr[i] = arr[i] / divisor;
                }
            }
            if (divisible) { lcm *= divisor; }
            else           { divisor++;  }
            if (count == arr.length) { return lcm; }
        }
    }

    private record Node(String left, String right) {}
}

File: 2024/src/Day08.cs


using System.Security.Cryptography.X509Certificates;

/**
 * @author Zachary Cockshutt
 * @since  2024-12-07
 */
public class Day08 : Day
{
    public override string PartOne()
    {
        var map = this.Input().ToCharMatrix();
        var positions = FindPositions(map);
        var antinodes = new HashSet<(int, int)>();
        foreach (var set in positions)
            foreach (var p1 in set.Value)
                foreach (var p2 in set.Value)
                {
                    if (p1 == p2) { continue; }
                    var p3 = (2*p1.x-p2.x, 2*p1.y-p2.y);
                    if (IsBounded(p3, map.Length)) { antinodes.Add(p3); }
                }
        return antinodes.Count.ToString();
    }

    public override string PartTwo()
    {
        var map = this.Input().ToCharMatrix();
        var positions = FindPositions(map);
        var antinodes = new HashSet<(int, int)>();
        foreach (var set in positions)
            for (int i = 0; i < set.Value.Count; i++)
                for (int j = 0; j < set.Value.Count; j++)
                {
                    var p1 = set.Value[i];
                    var p2 = set.Value[j];
                    if (p1 == p2) { continue; }
                    do
                    {
                        antinodes.Add(p2);
                        var p3 = (2*p1.x-p2.x, 2*p1.y-p2.y);
                        p2 = p1;
                        p1 = p3;
                    }
                    while (IsBounded(p2, map.Length));
                }
        return antinodes.Count.ToString();
    }

    private static Dictionary<char, List<(int x, int y)>> FindPositions(char[][] map)
    {
        int n = map.Length;
        var positions = new Dictionary<char, List<(int x, int y)>>();
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
            {
                char id = map[i][j];
                if (positions.TryAdd(id, [(i,j)])) { continue; }
                if (id != '.') { positions[id].Add((i,j)); }
            }
        return positions;
    }

    private static bool IsBounded(
        (int x, int y) pos, int n)
    {
        return pos.x >= 0 && pos.x < n
            && pos.y >= 0 && pos.y < n;
    }
}
